Reorganize the tutorials, take 2

melissawm opened the pull request:
Also:

* Consolidate debugging content
* Add note about flags and checkify to debug intro summary
* Update freshness

Finally, after feedback:
* Move parallel computations forward in Advanced Guides
* Add sharded_computations section back to tutorials.

Happy to address any other comments.

Thanks!

melissawm committed:
Reorganize the tutorials

Also:

* Consolidate debugging content
* Add note about flags and checkify to debug intro summary
* Update freshness
* Move parallel computation sections forward
* Add Training Cookbook to Advanced Guides
* Rename tutorials to jax-101
* Reorder tutorials
* Move Pytrees to advanced guides
* Adds redirects for deleted/moved pages
* Move sharded-computation to after random-numbers

melissawm commented:
cc @emilyfertig @mattjj

emilyfertig commented on docs/advanced_guides.rst:
```diff
@@ -0,0 +1,113 @@
+   jax-primitives
+   jaxpr
+
+.. toctree::
```
Could you move "Non-functional programming" to below "Performance benchmarking and profiling"?

emilyfertig commented on docs/tutorials.rst:
```diff
@@ -3,25 +3,20 @@
 Tutorials
 =========
 
+These tutorials cover basic usage of JAX and its features, including some of the
```
The new tutorial ordering is more of a sequential flow. Could you update the text to this?

These tutorials cover basic usage of JAX and its features, including some of the
internal mechanisms that make JAX work. They start with the fundamentals and are meant to be read sequentially. For more in-depth discussions of JAX's
design and implementation, in no particular order, see the :doc:`advanced guides <advanced_guides>`.

emilyfertig reviewed (changes requested):
Thank you!

emilyfertig reviewed (changes requested):
Just one more thing I missed, after that looks good! I'll give the JAX core folks a heads up in case they want to review.

jakevdp commented on docs/conf.py:
```diff
@@ -383,7 +383,7 @@ def linkcode_resolve(domain, info):
   "jax-101/08-pjit.rst": "sharded-computation.md",
-  "jax-101/index.rst": "tutorials.rst",
+  "jax-101/index.rst": "jax-101.rst",
   "notebooks/external_callbacks.md": "external-callbacks.md",
```
Any newly deleted sources should be added to this mapping to prevent broken hyperlinks (e.g. map `advanced_guide` to `advanced_guides`, and `tutorials` and `user_guides` should be mapped to whatever is replacing them).

jakevdp replied:
The goal is to create redirects in case there are existing URLs out there that point to the JAX docs. Each time we rename or remove a file, it would otherwise break those links.

melissawm replied:
Thanks! I have now added a redirect for "user_guides.rst" to point now to "advanced_guides.rst" since most of the User Guides were moved there. If that's not correct, please let me know.

emilyfertig commented on docs/jax-101.rst:
```diff
@@ -0,0 +1,21 @@
+   random-numbers
+   control-flow
+   stateful-computations
+   sharded-computation
```
One more thing I missed (thanks to Yash for spotting). Could you move this to after `random-numbers`? `sharded-computation` is "Intro to Parallel programming" in the doc.

emilyfertig commented:
Hi @melissawm -- this looks good to everyone and is ready to merge but https://github.com/jax-ml/jax/pull/31190 just went in and is causing a conflict, could you please resolve that, and squash to a single commit?

melissawm commented:
Done. One note is that initially I had removed the debugging/index file as most of its content was already in the debugging.md file. However, I see that recently more content was added so I added it back to the ToC, as this will need a second pass before removing.
